Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,629 @@
+                  " already exists")
+        return []
+    if not unsafe_hash:
+        if eq and not frozen:
+            return [Nodes.SingleAssignmentNode(node.pos,
+                                        lhs = ExprNodes.NameNode(node.pos, name=EncodedString("__hash__")),
+                                        rhs = ExprNodes.NoneNode(node.pos))]
+        if not eq:
+            return []
+
+    names = [ name for name, field in fields.items()
+                if (not field.is_initvar and
+                    (field.compare.value if field.hash.value is None else field.hash.value)) ]
+    if not names:
+        return []  # nothing to hash
+
+    # make a tuple of the hashes
+    tpl = u", ".join(u"hash(self.%s)" % name for name in names )
+
+    # if we're here we want to generate a hash
+    code_lines = u"""def __hash__(self):
+    return hash((%s))
+""" % tpl
+    code_tree = TreeFragment(code_lines,
+                              level='c_class', pipeline=[NormalizeTree(None)]
+                              ).substitute({})
+    return code_tree.stats
+
+
+class GetTypeNode(ExprNodes.ExprNode):
+    # Tries to return a pytype_type if possible. However contains
+    # some fallback provision if it turns out not to resolve to a Python object
+    # Initialize with "entry"
+
+    subexprs = []
+
+    def __init__(self, entry):
+        super(GetTypeNode, self).__init__(entry.pos, entry=entry)
+
+    def analyse_types(self, env):
+        type = self.entry.type
+
+        if type.is_extension_type or type.is_builtin_type:
+            return ExprNodes.RawCNameExprNode(self.pos, Builtin.type_type,
+                                                type.typeptr_cname).analyse_types(env)
+        else:
+            names = None
+            py_name = type.py_type_name()
+            # int types can return "(int, long)"
+            if py_name:
+                names = py_name.split(",")
+                names = [ n.strip("() ") for n in names ]
+            if names:
+                for name in names:
+                    name = EncodedString(name)
+                    nn = ExprNodes.NameNode(self.pos, name=name)
+                    # try to set the entry now to prevent the user accidentally shadowing
+                    # the name
+                    nn.entry = env.builtin_scope().lookup(name)
+                    if not nn.entry:
+                        try:
+                            nn.entry = env.declare_builtin(name, self.pos)
+                        except:
+                            pass  # not convinced a failure means much
+                    if nn.entry:
+                        return nn.analyse_types(env)
+
+        # otherwise we're left to return a string
+        s = self.entry.pep563_annotation
+        if not s:
+            s = self.entry.type.declaration_code("", for_display=1)
+        return ExprNodes.StringNode(self.pos, value=s).analyse_types(env)
+
+class DontAnalyseDeclarationsNode(ExprNodes.ExprNode):
```

Comment:
There must be a better way to do this. If it's about lambdas, then why can't they detect themselves whether they actually need to do something or not? Or maybe the transformation that uses this runs too late and should be executed before (or as part of) the analysis?

Reply:
I've added a guard to allow analyse_declarations be run twice on lambdas and generator expressions. I don't think it's possible to move the dataclass generation any earlier - it relies on the analyse_declarations stage having been run so that it know the types of things